File: cpp-implementation/src/products/pcgames/Strategy.cpp
#include "Product.ih"

// Constructor implementation for strategy games
Strategy::Strategy(int id, string name, double price, int volume, string genre, 
                   int ageRating, string theme, string supplier)
    : PC_Games(id, name, price, volume, genre, ageRating, supplier), 
      theme(theme) {}

// Calculates the discount depending on the current month and the packaged quantity
double Strategy::calculateDiscount(int volume, int month) const {
    if (month == 12) {
        return price * volume * 0.25;
    }
    return 0;
}

// Displays all the information
string Strategy::displayProductInfo() const {
    ostringstream info;

    info << "ID: " << id << "\nnGame: " << name << "\nSupplier: " << getSupplier()
    << "\nGenre: " << genre << "\nMinimum Age: " << ageRating
    << "\nTheme: " << theme << "\nnPrice: €" << price
    << "\nVolume: " << volume << " units\n\n";

    return info.str();

}

File: java-implementation/src/main/java/com/inventory/products/pcgames/Strategy.java
package com.inventory.products.pcgames;

// Represents a concrete subclass of PC_Games
public class Strategy extends PC_Games {
    private String theme;

    // Constructs an instance with specified details with base class attributes included
    public Strategy(int id, String name, double price, int volume, String genre, int ageRating, String theme, String supplier) {
        super(id, name, price, volume, genre, ageRating, supplier);
        this.theme = theme;
    }

    // Calculates the discount based on volume and month
    @Override
    public double calculateDiscount(int volume, int month) {
        if (month == 12) {
            return price * volume * 0.25;
        }
        return 0;
    }

    // Provides a string representation the details
    @Override
    public String displayProductInfo() {
        return "ID: " + id + "\nGame: " + name + "\nGenre: " + genre
               + "\nMinimum Age: " + ageRating + "\nTheme: " + theme
               + "\nPrice: €" + price + "\nVolume: " + volume + " units\nSupplier: " + supplier + "\n\n";
    }
}
